Tests/SonarScanner.MSBuild.PreProcessor.Test/SonarCloudWebServerTest.cs

```diff
@@ -139,233 +140,233 @@ public void DownloadProperties_Success()
+        Action act = () => _ = sut.DownloadProperties(null, null).Result;
+
+        act.Should().Throw<ArgumentNullException>().And.ParamName.Should().Be("projectKey");
+    }
+
+    [TestMethod]
+    public async Task IsServerLicenseValid_AlwaysTrue()
+    {
+        var isValid = await sut.IsServerLicenseValid();
+
+        isValid.Should().BeTrue();
+        logger.AssertDebugMessageExists("SonarCloud detected, skipping license check.");
+    }
+
+    [TestMethod]
+    public async Task DownloadCache_NullArgument()
+    {
+        (await sut.Invoking(x => x.DownloadCache(null)).Should().ThrowAsync<ArgumentNullException>()).And.ParamName.Should().Be("localSettings");
+    }
+
+    [TestMethod]
+    [DataRow("", "", "", "Incremental PR analysis: ProjectKey parameter was not provided.")]
+    [DataRow("project", "", "", "Incremental PR analysis: Base branch parameter was not provided.")]
+    [DataRow("project", "branch", "", "Incremental PR analysis: Token parameter was not provided.")]
+    [DataRow("project", "branch", "token", "Incremental PR analysis: CacheBaseUrl was not successfully retrieved.")]
+    public async Task DownloadCache_InvalidArguments(string projectKey, string branch, string token, string infoMessage)
+    {
+        sut = new SonarCloudWebServer(MockIDownloader(), version, logger, Organization, httpTimeout);
+        var localSettings = CreateLocalSettings(projectKey, branch, Organization, token);
+
+        var res = await sut.DownloadCache(localSettings);
+
+        res.Should().BeEmpty();
+        logger.AssertSingleInfoMessageExists(infoMessage);
+    }
+
+    [DataTestMethod]
+    [DataRow("Jenkins", "ghprbTargetBranch")]
+    [DataRow("Jenkins", "gitlabTargetBranch")]
+    [DataRow("Jenkins", "BITBUCKET_TARGET_BRANCH")]
+    [DataRow("GitHub Actions", "GITHUB_BASE_REF")]
+    [DataRow("GitLab", "CI_MERGE_REQUEST_TARGET_BRANCH_NAME")]
+    [DataRow("BitBucket Pipelines", "BITBUCKET_PR_DESTINATION_BRANCH")]
+    public async Task DownloadCache_AutomaticallyDeduceBaseBranch(string provider, string variableName)
+    {
+        using var environment = new EnvironmentVariableScope().SetVariable(variableName, "branch-42");
+        const string organization = "org42";
+        using var stream = new MemoryStream();
+        var handler = MockHttpHandler("http://myhost:222/v1/sensor_cache/prepare_read?organization=org42&project=project-key&branch=branch-42", "https://www.ephemeralUrl.com", stream);
+        sut = new SonarCloudWebServer(MockIDownloader("http://myhost:222"), version, logger, organization, httpTimeout, handler);
+        var localSettings = CreateLocalSettings(ProjectKey, null, organization, Token);
+
+        await sut.DownloadCache(localSettings);
+
+        logger.AssertInfoMessageExists($"Incremental PR analysis: Automatically detected base branch 'branch-42' from CI Provider '{provider}'.");
+        handler.Received().Protected("SendAsync", Arg.Any<HttpRequestMessage>(), Arg.Any<CancellationToken>());
```

Comment:
This is too hacky. Create `MockHttpHandler` class instead, that will inherit the abstract `HttpHandler` and implement the logic there.

Reply:
Why? It's actually very similar to what moq was doing before, I think it's slick and aligned with the NSubstitute invocations "style"